flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java

```diff
@@ -0,0 +1,172 @@
+    if (!startSnapshotOptional.isPresent()) {
+      return new ContinuousEnumerationResult(Collections.emptyList(), null,
+          new IcebergEnumeratorPosition(null, null));
+    }
+
+    Snapshot startSnapshot = startSnapshotOptional.get();
+    LOG.info("Get starting snapshot id {} based on strategy {}",
+        startSnapshot.snapshotId(), scanContext.startingStrategy());
+    List<IcebergSourceSplit> splits;
+    IcebergEnumeratorPosition toPosition;
+    if (scanContext.startingStrategy() == StreamingStartingStrategy.TABLE_SCAN_THEN_INCREMENTAL) {
+      // do a full table scan first
+      splits = FlinkSplitPlanner.planIcebergSourceSplits(table, scanContext, workerPool);
+      LOG.info("Discovered {} splits from initial full table scan with snapshot Id {}",
+          splits.size(), startSnapshot.snapshotId());
+      toPosition = new IcebergEnumeratorPosition(startSnapshot.snapshotId(), startSnapshot.timestampMillis());
+    } else {
+      splits = Collections.emptyList();
+      // Use parentId to achieve the inclusive behavior.
+      // Note that it is fine if parentId is null.
+      Long parentSnapshotId = startSnapshot.parentId();
+      Long parentSnapshotTimestampMs = null;
+      if (parentSnapshotId != null) {
+        parentSnapshotTimestampMs = table.snapshot(parentSnapshotId).timestampMillis();
+      }
+
+      toPosition = new IcebergEnumeratorPosition(parentSnapshotId, parentSnapshotTimestampMs);
+      LOG.info("Start incremental scan with start snapshot (inclusive): id = {}, timestamp = {}",
+          startSnapshot.snapshotId(), startSnapshot.timestampMillis());
+    }
+
+    return new ContinuousEnumerationResult(splits, null, toPosition);
+  }
+
+  /**
+   * Optional is used because table may be empty and has no snapshot
+   */
+  @VisibleForTesting
+  static Optional<Snapshot> getStartSnapshot(Table table, ScanContext scanContext) {
+    switch (scanContext.startingStrategy()) {
+      case TABLE_SCAN_THEN_INCREMENTAL:
+      case INCREMENTAL_FROM_LATEST_SNAPSHOT:
+        return Optional.ofNullable(table.currentSnapshot());
+      case INCREMENTAL_FROM_EARLIEST_SNAPSHOT:
+        return Optional.ofNullable(SnapshotUtil.oldestAncestor(table));
+      case INCREMENTAL_FROM_SNAPSHOT_ID:
+        Snapshot matchedSnapshotById = table.snapshot(scanContext.startSnapshotId());
+        if (matchedSnapshotById != null) {
+          return Optional.of(matchedSnapshotById);
+        } else {
+          throw new IllegalArgumentException(
+              "Start snapshot id not found in history: " + scanContext.startSnapshotId());
+        }
+      case INCREMENTAL_FROM_SNAPSHOT_TIMESTAMP:
+        // snapshotIdAsOfTime returns valid snapshot id only.
+        // it throws IllegalArgumentException when a matching snapshot not found.
+        long snapshotIdAsOfTime = SnapshotUtil.snapshotIdAsOfTime(table, scanContext.startSnapshotTimestamp());
```

Comment:
This is going to return the snapshot that was current as of some time:
* t0: s1 is committed
* t1
* t2: s2 is committed

`snapshotAsOfTime(t1)` will return s1. And it looks like this is going to return s1. Is that going to be exclusive or inclusive? I think it should be exclusive unless t0 == t1.

Reply:
I think to make this consistent, you should do the check here. If this snapshot's timestamp matches, return it for inclusive start. If not, then return the next snapshot for the inclusive start.